File: python/src/libghidra/headless.py
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: LicenseRef-Human-Origin-Source-1.0
#
# This file is licensed under the Human-Origin Source License v1.0.
# See LICENSE.
"""Launch headless Ghidra and return a connected GhidraClient."""

from collections import deque
import os
import shutil
import socket
import subprocess
import tempfile
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable, List, Optional
from urllib.parse import urlparse

from .client import ClientOptions, GhidraClient

READY_BANNER = "LIBGHIDRA_HEADLESS_READY"


@dataclass
class HeadlessProjectOptions:
    """Options for launching a project-scoped headless Ghidra RPC host."""

    ghidra_dir: str = ""
    port: int = 18080
    bind: str = "127.0.0.1"          # bind address for the headless server
    project_dir: str = ""
    project_name: str = "HeadlessProject"
    shutdown: str = "save"            # "save", "discard", or "none"
    auth_token: str = ""              # bearer auth token
    max_runtime_seconds: int = 0      # 0 = no limit (forwarded as max_runtime_ms)
    bind_attempts: int = 1
    startup_timeout: float = 300.0
    read_timeout: float = 300.0
    script_dir: str = ""              # override script dir (empty = auto-detect)
    extra_script_args: List[str] = field(default_factory=list)
    extra_headless_args: List[str] = field(default_factory=list)
    on_output: Optional[Callable[[str], None]] = None


class HeadlessClient:
    """A GhidraClient backed by a headless Ghidra process.

    Use as a context manager or call close() when done.
    """

    def __init__(self, client: GhidraClient, proc: subprocess.Popen,
                 project_dir: Path, owns_project_dir: bool,
                 base_url: str,
                 shutdown_policy: int,
                 on_output: Optional[Callable[[str], None]] = None,
                 output_thread: Optional[threading.Thread] = None):
        self._client = client
        self._proc = proc
        self._project_dir = project_dir
        self._owns_project_dir = owns_project_dir
        self._base_url = base_url
        self._shutdown_policy = shutdown_policy
        self._on_output = on_output
        self._output_thread = output_thread

    @property
    def client(self) -> GhidraClient:
        return self._client

    @property
    def base_url(self) -> str:
        return self._base_url

    @property
    def process(self) -> subprocess.Popen:
        return self._proc

    @property
    def project_dir(self) -> Path:
        return self._project_dir

    def detach(self) -> None:
        """Release the process handle without killing it.

        After detach(), close() and __exit__ become no-ops for the process.
        """
        self._proc = None
        self._owns_project_dir = False

    def close(self, save: bool | None = None) -> int:
        """Shut down the Ghidra host and wait for the process to exit."""
        if self._proc is None:
            return 0

        from .models import ShutdownPolicy
        shutdown_requested = False
        try:
            if save is None:
                policy = ShutdownPolicy(self._shutdown_policy)
            else:
                policy = ShutdownPolicy.SAVE if save else ShutdownPolicy.DISCARD
            self._client.shutdown(policy)
            shutdown_requested = True
        except Exception:
            pass

        if shutdown_requested:
            _wait_for_endpoint_closed(self._base_url, timeout=60.0)

        # Keep draining synchronously only for older callers without a drainer.
        if self._proc.stdout and self._output_thread is None:
            for line in self._proc.stdout:
                line = line.strip()
                if line and self._on_output:
                    self._on_output(line)

        try:
            self._proc.wait(timeout=60)
        except subprocess.TimeoutExpired:
            self._proc.kill()
            self._proc.wait(timeout=10)

        exit_code = self._proc.returncode
        if self._output_thread is not None:
            self._output_thread.join(timeout=5)

        if self._owns_project_dir:
            shutil.rmtree(self._project_dir, ignore_errors=True)

        return exit_code

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

    def __getattr__(self, name):
        """Delegate all GhidraClient methods directly."""
        return getattr(self._client, name)


def _find_launcher(ghidra_dir: Path) -> Path:
    names = (
        ("support/analyzeHeadless.bat", "support/analyzeHeadless")
        if os.name == "nt"
        else ("support/analyzeHeadless", "support/analyzeHeadless.bat")
    )
    for name in names:
        p = ghidra_dir / name
        if p.exists():
            return p
    raise FileNotFoundError(
        f"analyzeHeadless not found in {ghidra_dir}/support/")


def _find_script_dir(ghidra_dir: Path) -> Path:
    d = ghidra_dir / "Ghidra" / "Extensions" / "LibGhidraHost" / "ghidra_scripts"
    if not d.exists():
        raise FileNotFoundError(
            f"LibGhidraHost extension not installed at {d.parent}\n"
            "Install it first: gradle installExtension -PGHIDRA_INSTALL_DIR=<dist>")
    return d


def _emit_output_line(line: str, on_output: Optional[Callable[[str], None]]) -> None:
    line = line.strip()
    if line and on_output:
        on_output(line)


def _start_output_drain(
        proc: subprocess.Popen,
        on_output: Optional[Callable[[str], None]] = None) -> Optional[threading.Thread]:
    if proc.stdout is None:
        return None

    def drain() -> None:
        try:
            for line in proc.stdout:
                _emit_output_line(line, on_output)
        except Exception:
            pass

    thread = threading.Thread(
        target=drain,
        name="libghidra-headless-output",
        daemon=True)
    thread.start()
    return thread


def _shutdown_policy_value(name: str) -> int:
    from .models import ShutdownPolicy
    normalized = name.strip().lower()
    if normalized == "save":
        return int(ShutdownPolicy.SAVE)
    if normalized == "discard":
        return int(ShutdownPolicy.DISCARD)
    if normalized == "none":
        return int(ShutdownPolicy.NONE)
    return int(ShutdownPolicy.UNSPECIFIED)


def _wait_for_endpoint_closed(base_url: str, timeout: float) -> bool:
    parsed = urlparse(base_url)
    host = parsed.hostname or "127.0.0.1"
    port = parsed.port or (443 if parsed.scheme == "https" else 80)
    deadline = time.monotonic() + timeout

    while time.monotonic() < deadline:
        try:
            with socket.create_connection((host, port), timeout=0.5):
                pass
        except OSError:
            return True
        time.sleep(0.2)

    return False


def _append_output_line(
        tail: deque[str],
        line: str,
        on_output: Optional[Callable[[str], None]] = None) -> None:
    line = line.strip()
    if not line:
        return
    tail.append(line)
    if on_output:
        on_output(line)


def _drain_remaining_output(
        proc: subprocess.Popen,
        tail: deque[str],
        on_output: Optional[Callable[[str], None]] = None) -> None:
    if proc.stdout is None or proc.poll() is None:
        return
    for line in proc.stdout.readlines():
        _append_output_line(tail, line, on_output)


def _format_tail(tail: deque[str] | list[str]) -> str:
    if not tail:
        return ""
    lines = list(tail)
    diagnostic_terms = ("ERROR", "Exception", "BindException", "Address already")
    diagnostic_lines = [
        line for line in lines
        if any(term in line for term in diagnostic_terms)
    ][-12:]
    recent_lines = lines[-40:]
    selected: list[str] = []
    seen: set[str] = set()
    for line in [*diagnostic_lines, *recent_lines]:
        if line in seen:
            continue
        seen.add(line)
        selected.append(line)
    return "\nRecent Ghidra output:\n" + "\n".join(f"  {line}" for line in selected)


def launch_headless_project(opts: HeadlessProjectOptions) -> HeadlessClient:
    """Launch a project-scoped headless Ghidra host and return a client.

    >>> import libghidra as ghidra
    >>> with ghidra.launch_headless_project(ghidra.HeadlessProjectOptions(
    ...     ghidra_dir="/path/to/ghidra_dist",
    ... )) as h:
    ...     imported = h.import_program(ghidra.ImportProgramRequest(
    ...         source_path="/path/to/target.exe", analyze=True, overwrite=True))
    ...     h.open_program(ghidra.OpenProgramRequest(
    ...         program_path=imported.primary_program_path))
    """
    ghidra_dir = Path(opts.ghidra_dir).resolve()

    launcher = _find_launcher(ghidra_dir)

    if opts.script_dir:
        script_dir = Path(opts.script_dir)
    else:
        script_dir = _find_script_dir(ghidra_dir)

    owns_project_dir = not opts.project_dir
    project_dir = Path(opts.project_dir) if opts.project_dir else Path(
        tempfile.mkdtemp(prefix="ghidra_headless_"))
    project_dir.mkdir(parents=True, exist_ok=True)

    on_output = opts.on_output

    # Start analyzeHeadless with scripts but without -import/-process. Ghidra
    # creates/opens the project and runs the script with no active program;
    # ImportProgram/OpenProgram RPCs drive the project after READY.
    cmd = [
        str(launcher),
        str(project_dir), opts.project_name,
    ]
    cmd.extend(opts.extra_headless_args)
    cmd += [
        "-scriptPath", str(script_dir),
        "-postScript", "LibGhidraHeadlessServer.java",
        f"bind={opts.bind}",
        f"port={opts.port}",
        f"shutdown={opts.shutdown}",
    ]
    if opts.auth_token:
        cmd.append(f"auth={opts.auth_token}")
    if opts.max_runtime_seconds > 0:
        cmd.append(f"max_runtime_ms={opts.max_runtime_seconds * 1000}")
    if opts.bind_attempts > 1:
        cmd.append(f"bind_attempts={opts.bind_attempts}")
    cmd.extend(opts.extra_script_args)

    proc = subprocess.Popen(
        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
        text=True, bufsize=1)

    # Wait for LIBGHIDRA_HEADLESS_READY banner
    actual_port = opts.port
    deadline = time.monotonic() + opts.startup_timeout
    tail: deque[str] = deque(maxlen=200)
    try:
        while time.monotonic() < deadline:
            if proc.poll() is not None:
                _drain_remaining_output(proc, tail, on_output)
                raise RuntimeError(
                    f"Ghidra exited prematurely (code={proc.returncode})"
                    f"{_format_tail(tail)}")
            line = proc.stdout.readline()
            if not line:
                time.sleep(0.1)
                continue
            _append_output_line(tail, line, on_output)
            if READY_BANNER in line:
                for part in line.split():
                    if part.startswith("port="):
                        actual_port = int(part.split("=", 1)[1])
                break
        else:
            _drain_remaining_output(proc, tail, on_output)
            raise TimeoutError(
                f"Timed out after {opts.startup_timeout}s waiting for Ghidra"
                f"{_format_tail(tail)}")
    except BaseException:
        proc.terminate()
        try:
            proc.wait(timeout=10)
        except subprocess.TimeoutExpired:
            proc.kill()
        if owns_project_dir:
            shutil.rmtree(project_dir, ignore_errors=True)
        raise

    # Connect
    base_url = f"http://{opts.bind}:{actual_port}"
    client_opts = ClientOptions(
        base_url=base_url,
        read_timeout=opts.read_timeout,
    )
    if opts.auth_token:
        client_opts.auth_token = opts.auth_token
    client = GhidraClient(client_opts)

    output_thread = _start_output_drain(proc, on_output)
    return HeadlessClient(client, proc, project_dir, owns_project_dir,
                          base_url, _shutdown_policy_value(opts.shutdown),
                          on_output, output_thread)

File: rust/src/headless.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: LicenseRef-Human-Origin-Source-1.0
//
// This file is licensed under the Human-Origin Source License v1.0.
// See LICENSE.
//! Launch headless Ghidra and return a connected [`GhidraClient`].

use std::io::BufRead;
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdout, Command, Stdio};
use std::thread::JoinHandle;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use crate::client::{ClientOptions, GhidraClient};
use crate::error::{Error, ErrorCode};
use crate::models::ShutdownPolicy;

const READY_BANNER: &str = "LIBGHIDRA_HEADLESS_READY";

/// Canonicalize a path, stripping the `\\?\` prefix on Windows so that
/// Java (Ghidra) can parse it.
fn clean_canonicalize(p: &str) -> std::io::Result<PathBuf> {
    let canon = std::fs::canonicalize(p)?;
    if cfg!(windows) {
        let s = canon.to_string_lossy();
        if let Some(stripped) = s.strip_prefix(r"\\?\") {
            return Ok(PathBuf::from(stripped));
        }
    }
    Ok(canon)
}

/// A sink for a headless host's stdout lines (progress/log streaming).
pub type OutputCallback = Box<dyn Fn(&str) + Send>;

/// Options for launching a project-scoped headless Ghidra RPC host.
pub struct HeadlessProjectOptions {
    pub ghidra_dir: String,
    pub port: u16,
    /// Bind address for the headless server.
    pub bind: String,
    pub project_dir: String,
    pub project_name: String,
    /// Shutdown policy: "save", "discard", or "none".
    pub shutdown: String,
    /// Bearer auth token.
    pub auth_token: String,
    /// Max runtime in seconds (0 = no limit, forwarded as max_runtime_ms).
    pub max_runtime_seconds: u64,
    pub bind_attempts: u32,
    pub startup_timeout: Duration,
    pub read_timeout: Duration,
    pub script_dir: String,
    pub extra_script_args: Vec<String>,
    pub extra_headless_args: Vec<String>,
    pub on_output: Option<OutputCallback>,
}

impl Default for HeadlessProjectOptions {
    fn default() -> Self {
        Self {
            ghidra_dir: String::new(),
            port: 18080,
            bind: "127.0.0.1".to_string(),
            project_dir: String::new(),
            project_name: "HeadlessProject".to_string(),
            shutdown: "save".to_string(),
            auth_token: String::new(),
            max_runtime_seconds: 0,
            bind_attempts: 1,
            startup_timeout: Duration::from_secs(300),
            read_timeout: Duration::from_secs(300),
            script_dir: String::new(),
            extra_script_args: Vec::new(),
            extra_headless_args: Vec::new(),
            on_output: None,
        }
    }
}

/// A connected client backed by a headless Ghidra process.
pub struct HeadlessClient {
    client: GhidraClient,
    child: Option<Child>,
    output_thread: Option<JoinHandle<()>>,
    base_url: String,
    project_dir: PathBuf,
    owns_project_dir: bool,
}

impl HeadlessClient {
    /// The connected RPC client.
    pub fn client(&self) -> &GhidraClient {
        &self.client
    }

    /// The base URL the client is connected to.
    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    /// The project directory used by this headless session.
    pub fn project_dir(&self) -> &Path {
        &self.project_dir
    }

    /// Release the process handle without killing it.
    /// After detach(), Drop and close() become no-ops for the process.
    pub fn detach(&mut self) {
        let _ = self.child.take();
        let _ = self.output_thread.take();
        self.owns_project_dir = false;
    }

    /// Shut down the host, wait for exit, clean up. Returns the exit code.
    pub fn close(&mut self, save: bool) -> i32 {
        let policy = if save {
            ShutdownPolicy::Save
        } else {
            ShutdownPolicy::Discard
        };
        let _ = self.client.shutdown(policy);

        let code = match self.child.take() {
            Some(mut c) => match c.wait() {
                Ok(s) => s.code().unwrap_or(-1),
                Err(_) => -1,
            },
            None => 0,
        };

        if let Some(output_thread) = self.output_thread.take() {
            let _ = output_thread.join();
        }

        if self.owns_project_dir {
            let _ = std::fs::remove_dir_all(&self.project_dir);
        }
        code
    }
}

impl std::ops::Deref for HeadlessClient {
    type Target = GhidraClient;
    fn deref(&self) -> &GhidraClient {
        &self.client
    }
}

impl std::ops::DerefMut for HeadlessClient {
    fn deref_mut(&mut self) -> &mut GhidraClient {
        &mut self.client
    }
}

impl Drop for HeadlessClient {
    fn drop(&mut self) {
        if let Some(ref mut child) = self.child {
            let _ = child.kill();
            let _ = child.wait();
        }
        if let Some(output_thread) = self.output_thread.take() {
            let _ = output_thread.join();
        }
        if self.owns_project_dir {
            let _ = std::fs::remove_dir_all(&self.project_dir);
        }
    }
}

fn emit_output_line(line: &str, on_output: Option<&(dyn Fn(&str) + Send)>) {
    let trimmed = line.trim();
    if !trimmed.is_empty() {
        if let Some(cb) = on_output {
            cb(trimmed);
        }
    }
}

fn start_output_drainer(
    stdout: ChildStdout,
    on_output: Option<OutputCallback>,
) -> std::io::Result<JoinHandle<()>> {
    std::thread::Builder::new()
        .name("libghidra-headless-output".to_string())
        .spawn(move || {
            let reader = std::io::BufReader::new(stdout);
            for line in reader.lines().map_while(|l| l.ok()) {
                emit_output_line(&line, on_output.as_deref());
            }
        })
}

fn find_launcher(ghidra_dir: &Path) -> Result<PathBuf, Error> {
    let candidates = if cfg!(windows) {
        vec![ghidra_dir.join("support").join("analyzeHeadless.bat")]
    } else {
        vec![ghidra_dir.join("support").join("analyzeHeadless")]
    };
    for c in &candidates {
        if c.exists() {
            return Ok(c.clone());
        }
    }
    Err(Error::new(
        ErrorCode::NotFound,
        format!(
            "analyzeHeadless not found in {}/support/",
            ghidra_dir.display()
        ),
    ))
}

fn find_script_dir(ghidra_dir: &Path) -> Result<PathBuf, Error> {
    let d = ghidra_dir
        .join("Ghidra")
        .join("Extensions")
        .join("LibGhidraHost")
        .join("ghidra_scripts");
    if d.exists() {
        Ok(d)
    } else {
        Err(Error::new(
            ErrorCode::NotFound,
            format!(
                "LibGhidraHost extension not installed at {}",
                d.parent().unwrap_or(&d).display()
            ),
        ))
    }
}

/// Launch a project-scoped headless Ghidra host and return a connected client.
///
/// ```no_run
/// use libghidra as ghidra;
/// let mut h = ghidra::launch_headless_project(ghidra::HeadlessProjectOptions {
///     ghidra_dir: "/path/to/ghidra_dist".into(),
///     ..Default::default()
/// }).unwrap();
/// h.close(true);
/// ```
pub fn launch_headless_project(opts: HeadlessProjectOptions) -> Result<HeadlessClient, Error> {
    let ghidra_dir = clean_canonicalize(&opts.ghidra_dir).map_err(|e| {
        Error::new(
            ErrorCode::NotFound,
            format!("Ghidra dir not found: {}: {e}", opts.ghidra_dir),
        )
    })?;

    let launcher = find_launcher(&ghidra_dir)?;
    let script_dir = if opts.script_dir.is_empty() {
        find_script_dir(&ghidra_dir)?
    } else {
        PathBuf::from(&opts.script_dir)
    };

    let owns_project_dir = opts.project_dir.is_empty();
    let project_dir = if owns_project_dir {
        let suffix = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0);
        std::env::temp_dir().join(format!("ghidra_headless_rust_{suffix}"))
    } else {
        PathBuf::from(&opts.project_dir)
    };
    std::fs::create_dir_all(&project_dir).map_err(|e| {
        Error::new(
            ErrorCode::TransportError,
            format!(
                "Failed to create headless project dir {}: {e}",
                project_dir.display()
            ),
        )
    })?;

    // Build command. No -import and no -process: Ghidra creates/opens
    // the project, runs the server script without an active program, and
    // callers drive ImportProgram/OpenProgram explicitly over RPC.
    let mut cmd = Command::new(&launcher);
    let _ = cmd.arg(&project_dir).arg(&opts.project_name);
    for arg in &opts.extra_headless_args {
        let _ = cmd.arg(arg);
    }
    let _ = cmd
        .arg("-scriptPath")
        .arg(&script_dir)
        .arg("-postScript")
        .arg("LibGhidraHeadlessServer.java")
        .arg(format!("bind={}", opts.bind))
        .arg(format!("port={}", opts.port))
        .arg(format!("shutdown={}", opts.shutdown));
    if !opts.auth_token.is_empty() {
        let _ = cmd.arg(format!("auth={}", opts.auth_token));
    }
    if opts.max_runtime_seconds > 0 {
        let _ = cmd.arg(format!(
            "max_runtime_ms={}",
            opts.max_runtime_seconds as u128 * 1000
        ));
    }
    if opts.bind_attempts > 1 {
        let _ = cmd.arg(format!("bind_attempts={}", opts.bind_attempts));
    }
    for arg in &opts.extra_script_args {
        let _ = cmd.arg(arg);
    }
    let _ = cmd.stdout(Stdio::piped()).stderr(Stdio::inherit());

    let mut child = cmd.spawn().map_err(|e| {
        Error::new(
            ErrorCode::TransportError,
            format!("Failed to launch analyzeHeadless: {e}"),
        )
    })?;

    // Merge stderr into stdout reader by taking stdout
    // (stderr is separate but we'll read stdout which has the banner)
    let stdout = child
        .stdout
        .take()
        .ok_or_else(|| Error::new(ErrorCode::TransportError, "Failed to capture stdout"))?;

    // Wait for LIBGHIDRA_HEADLESS_READY banner using read_line so we
    // keep ownership of the BufReader and can return stdout to the child.
    let deadline = Instant::now() + opts.startup_timeout;
    let mut reader = std::io::BufReader::new(stdout);
    let mut actual_port = opts.port;
    let mut found = false;
    let mut line_buf = String::new();

    while Instant::now() < deadline {
        line_buf.clear();
        match reader.read_line(&mut line_buf) {
            Ok(0) => {
                // EOF — process exited
                let code = child.wait().map(|s| s.code().unwrap_or(-1)).unwrap_or(-1);
                return Err(Error::new(
                    ErrorCode::TransportError,
                    format!("Ghidra exited prematurely (code={code})"),
                ));
            }
            Ok(_) => {
                let trimmed = line_buf.trim().to_string();
                emit_output_line(&trimmed, opts.on_output.as_deref());
                if trimmed.contains(READY_BANNER) {
                    for part in trimmed.split_whitespace() {
                        if let Some(val) = part.strip_prefix("port=") {
                            if let Ok(p) = val.parse::<u16>() {
                                actual_port = p;
                            }
                        }
                    }
                    found = true;
                    break;
                }
            }
            Err(_) => {
                std::thread::sleep(Duration::from_millis(100));
            }
        }
    }

    if !found {
        let _ = child.kill();
        let _ = child.wait();
        if owns_project_dir {
            let _ = std::fs::remove_dir_all(&project_dir);
        }
        return Err(Error::new(
            ErrorCode::Timeout,
            format!(
                "Timed out after {}s waiting for Ghidra to start",
                opts.startup_timeout.as_secs()
            ),
        ));
    }

    let output_thread = match start_output_drainer(reader.into_inner(), opts.on_output) {
        Ok(thread) => thread,
        Err(e) => {
            let _ = child.kill();
            let _ = child.wait();
            if owns_project_dir {
                let _ = std::fs::remove_dir_all(&project_dir);
            }
            return Err(Error::new(
                ErrorCode::TransportError,
                format!("Failed to start headless output drainer: {e}"),
            ));
        }
    };

    // Connect
    let base_url = format!("http://{}:{actual_port}", opts.bind);
    let client = GhidraClient::new(ClientOptions {
        base_url: base_url.clone(),
        auth_token: opts.auth_token.clone(),
        read_timeout: opts.read_timeout,
        ..Default::default()
    });

    Ok(HeadlessClient {
        client,
        child: Some(child),
        output_thread: Some(output_thread),
        base_url,
        project_dir,
        owns_project_dir,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::process::{Command, Stdio};
    use std::sync::{
        atomic::{AtomicUsize, Ordering},
        Arc,
    };

    #[test]
    fn generic_headless_options_default_to_no_action() {
        assert_eq!(HeadlessProjectOptions::default().shutdown, "save");
    }

    #[test]
    fn output_drainer_consumes_child_stdout_until_eof() {
        const LINE_COUNT: usize = 10_000;

        let mut cmd = output_flood_command(LINE_COUNT);
        let _ = cmd.stdout(Stdio::piped()).stderr(Stdio::null());
        let mut child = cmd.spawn().expect("spawn stdout writer");
        let stdout = child.stdout.take().expect("capture child stdout");

        let seen = Arc::new(AtomicUsize::new(0));
        let seen_by_callback = Arc::clone(&seen);
        let output_thread = start_output_drainer(
            stdout,
            Some(Box::new(move |_| {
                let _ = seen_by_callback.fetch_add(1, Ordering::Relaxed);
            })),
        )
        .expect("start output drainer");

        let status = child.wait().expect("wait for stdout writer");
        assert!(status.success(), "stdout writer failed: {status:?}");
        output_thread.join().expect("join output drainer");
        assert_eq!(seen.load(Ordering::Relaxed), LINE_COUNT);
    }

    #[cfg(windows)]
    fn output_flood_command(line_count: usize) -> Command {
        let mut cmd = Command::new("cmd");
        let _ = cmd
            .arg("/C")
            .arg(format!("for /L %i in (1,1,{line_count}) do @echo line%i"));
        cmd
    }

    #[cfg(not(windows))]
    fn output_flood_command(line_count: usize) -> Command {
        let mut cmd = Command::new("sh");
        let _ = cmd.arg("-c").arg(format!(
            "i=1; while [ $i -le {line_count} ]; do echo line$i; i=$((i+1)); done"
        ));
        cmd
    }
}
